Sources/Fluent/Entity/Entity.swift

```diff
@@ -37,45 +39,29 @@ public protocol Entity: Preparation, NodeConvertible {
     */
     var id: Node? { get set }
 
-    /**
-        Whether or not entity was retrieved from database.
-        
-        This value shouldn't be interacted w/ external users 
-        w/o explicit knowledge.
-     
-        General implementation should just be `var exists = false`
-    */
-    var exists: Bool { get set }
+    /// The type of identifier this model uses.
+    /// ex: INT, UUID, etc
+    static var idType: IdentifierType { get }
 
-    /**
-        Called before the entity will be created.
-    */
-    func willCreate()
+    /// Called before the entity will be created.
+    func willCreate() throws
 
-    /**
-        Called after the entity has been created.
-    */
-    func didCreate()
+    /// Called after the entity has been created.
+    func didCreate() throws
 
-    /**
-        Called before the entity will be updated.
-    */
-    func willUpdate()
+    /// Called before the entity will be updated.
+    func willUpdate() throws
 
-    /**
-        Called after the entity has been updated.
-    */
-    func didUpdate()
+    /// Called after the entity has been updated.
+    func didUpdate() throws
```

Comment:
That's a good point. Throwing after a did makes it unclear what will happen next.